File: rust/src/colorprint.rs
//! # Colorprint
//! Interface for coloured terminal printing.
//!
//! Akseli Lukkarila
//! 2019-2026

use colored::Colorize;

/// Print bold text with the given format string and arguments.
#[macro_export]
macro_rules! print_bold {
    ($($arg:tt)*) => {{
        use colored::Colorize;
        println!("{}", format!($($arg)*).bold())
    }};
}

/// Print text in green.
#[macro_export]
macro_rules! print_green {
    ($($arg:tt)*) => {{
        use colored::Colorize;
        println!("{}", format!($($arg)*).green())
    }};
}

/// Print bold text in green.
#[macro_export]
macro_rules! print_green_bold {
    ($($arg:tt)*) => {{
        use colored::Colorize;
        println!("{}", format!($($arg)*).green().bold())
    }};
}

/// Print text in yellow.
#[macro_export]
macro_rules! print_yellow {
    ($($arg:tt)*) => {{
        use colored::Colorize;
        println!("{}", format!($($arg)*).yellow())
    }};
}

/// Print bold text in yellow.
#[macro_export]
macro_rules! print_yellow_bold {
    ($($arg:tt)*) => {{
        use colored::Colorize;
        println!("{}", format!($($arg)*).yellow().bold())
    }};
}

/// Print text in red.
#[macro_export]
macro_rules! print_red {
    ($($arg:tt)*) => {{
        use colored::Colorize;
        println!("{}", format!($($arg)*).red())
    }};
}

/// Print bold text in red.
#[macro_export]
macro_rules! print_red_bold {
    ($($arg:tt)*) => {{
        use colored::Colorize;
        println!("{}", format!($($arg)*).red().bold())
    }};
}

/// Print error message with red colour.
pub fn print_error(message: &str) {
    let (indent, text) = split_leading_whitespace(message);
    println!("{}", format!("{indent}Error: {text}").red());
}

/// Print warning message with yellow colour.
pub fn print_warn(message: &str) {
    let (indent, text) = split_leading_whitespace(message);
    println!("{}", format!("{indent}Warning: {text}").yellow());
}

/// Split a string into the leading whitespace and the rest of the string.
fn split_leading_whitespace(message: &str) -> (&str, &str) {
    let indent_size = message.chars().take_while(|c| c.is_whitespace()).count();
    message.split_at(indent_size)
}

File: typescript/src/colorprint.ts
//==========================================================
// Colorprint
// Interface for coloured terminal printing
// Akseli Lukkarila
// 2019-2026
//==========================================================

import chalk from 'chalk';

/** Print text with bold style. */
export function printBold(text: string): void {
    console.log(chalk.bold(text));
}

/** Print text with green colour. */
export function printGreen(text: string): void {
    console.log(chalk.green(text));
}

/** Print text with green colour and bold style. */
export function printGreenBold(text: string): void {
    console.log(chalk.green.bold(text));
}

/** Print text with yellow colour. */
export function printYellow(text: string): void {
    console.log(chalk.yellow(text));
}

/** Print text with yellow colour and bold style. */
export function printYellowBold(text: string): void {
    console.log(chalk.yellow.bold(text));
}

/** Print text with red colour. */
export function printRed(text: string): void {
    console.log(chalk.red(text));
}

/** Print text with red colour and bold style. */
export function printRedBold(text: string): void {
    console.log(chalk.red.bold(text));
}

/** Print error message with red colour. */
export function printError(message: string): void {
    const [indent, text] = splitLeadingWhitespace(message);
    console.log(chalk.red(`${indent}Error: ${text}`));
}

/** Print warning message with yellow colour. */
export function printWarn(message: string): void {
    const [indent, text] = splitLeadingWhitespace(message);
    console.log(chalk.yellow(`${indent}Warning: ${text}`));
}

/** Split a string into the leading whitespace and the rest of the string. */
function splitLeadingWhitespace(message: string): [string, string] {
    const indentSize = message.length - message.trimStart().length;
    return [message.slice(0, indentSize), message.slice(indentSize)];
}
